File: auditlog-go/transaction.go
package audit

import (
	"context"
	"sync"
)

// Transaction is a one-owner, process-local helper for one Audit
// Transaction lifecycle: NEW -> START_IN_FLIGHT -> STARTED, then STARTED
// -> COMPLETE_IN_FLIGHT -> COMPLETED or STARTED -> ABORT_IN_FLIGHT ->
// ABORTED. It is NOT safe for concurrent use by multiple goroutines in
// the sense of issuing two overlapping Record/Complete/Abort calls —
// construct one Transaction per logical business operation and do not
// share it across goroutines that might call it concurrently. It does
// serialize itself defensively (a busy guard rejects an overlapping
// call rather than corrupting state), but that guard exists to fail
// loudly, not to make concurrent use a supported pattern. The main
// Client remains safe for concurrent use across many Transactions.
//
// A Transaction holds no durable state and cannot be resumed after a
// process restart. If the process crashes after Record succeeds but
// before Complete/Abort is called, the evidence honestly remains
// STARTED — see the README's "Transaction lifecycle" section and
// internal/testutils/testdata/contracts/sdk/v1/transactions/manifest.json's
// process_crash_recovery guidance.
type Transaction struct {
	mu     sync.Mutex
	client *Client
	state  TransactionState

	reserved           bool
	auditTransactionID string
	eventID            string
	startRecordID      string
	startPrepared      preparedTransactionStart

	terminalChosen   bool
	terminalKind     TransactionState // TransactionStateCompleted or TransactionStateAborted once chosen
	terminalPrepared preparedTransactionTerminal
}

// NewTransaction constructs a new, unstarted Transaction bound to c.
func (c *Client) NewTransaction() *Transaction {
	return newTransaction(c)
}

func newTransaction(c *Client) *Transaction {
	return &Transaction{client: c, state: TransactionStateNew}
}

func (t *Transaction) lock()   { t.mu.Lock() }
func (t *Transaction) unlock() { t.mu.Unlock() }

// State reports the Transaction's current state. Safe for concurrent use.
func (t *Transaction) State() TransactionState {
	t.lock()
	defer t.unlock()
	return t.state
}

// AuditTransactionID reports the reserved transaction ID, or "" before
// the first Record call has reserved one. A caller that needs to persist
// this ID for restart-recovery purposes (see the README) should read it
// only after Record has returned successfully.
func (t *Transaction) AuditTransactionID() string {
	t.lock()
	defer t.unlock()
	return t.auditTransactionID
}

// EventID reports the reserved event ID, or "" before the first Record
// call has reserved one.
func (t *Transaction) EventID() string {
	t.lock()
	defer t.unlock()
	return t.eventID
}

// StartRecordID reports the reserved start record ID, or "" before the
// first Record call has reserved one.
func (t *Transaction) StartRecordID() string {
	t.lock()
	defer t.unlock()
	return t.startRecordID
}

// newTransactionStateError reports an invalid call against the
// Transaction's current state (busy, wrong order, already terminal, or
// mutual exclusion). It returns *APIError (CodeValidationError,
// non-retryable) so callers branch on err.(*APIError) exactly the way
// they already do for every other validation failure in this module.
func newTransactionStateError(message string) error {
	return newValidationError(message)
}

// Record prepares input (generating and reserving audit_transaction_id/
// event_id/record_id on first call) and durably starts the transaction,
// retrying eligible transient failures with the exact same prepared
// bytes. It resolves for both ACCEPTED and DUPLICATE.
//
// Calling Record again after it already succeeded is rejected — a
// Transaction starts at most once. Calling Record again after a failed
// attempt is allowed and reuses the exact same reserved IDs and prepared
// bytes, so a caller-driven retry (beyond this call's own internal retry
// budget) remains idempotent.
func (t *Transaction) Record(ctx context.Context, input TransactionStartInput, opts RecordOptions) (TransactionStartReceipt, error) {
	t.lock()
	switch t.state {
	case TransactionStateStartInFlight, TransactionStateCompleteInFlight, TransactionStateAbortInFlight:
		t.unlock()
		return TransactionStartReceipt{}, newTransactionStateError("transaction: a call is already in flight; overlapping calls are not allowed")
	case TransactionStateStarted, TransactionStateCompleted, TransactionStateAborted:
		t.unlock()
		return TransactionStartReceipt{}, newTransactionStateError("transaction: Record can only succeed once per transaction; this transaction has already started")
	}
	// t.state == TransactionStateNew here.
	if !t.reserved {
		prepared, err := createTransactionStart(input, t.client.now, t.client.random)
		if err != nil {
			t.unlock()
			return TransactionStartReceipt{}, err
		}
		t.reserved = true
		t.auditTransactionID = prepared.AuditTransactionID
		t.eventID = prepared.EventID
		t.startRecordID = prepared.RecordID
		t.startPrepared = prepared
	}
	prepared := t.startPrepared
	t.state = TransactionStateStartInFlight
	t.unlock()

	receipt, err := t.client.sendTransactionStart(ctx, prepared, opts)

	t.lock()
	if err != nil {
		// Revert to NEW (not a terminal failure state) so a caller-driven
		// retry can call Record again and reuse the exact same reserved
		// IDs and bytes reserved above.
		t.state = TransactionStateNew
		t.unlock()
		return TransactionStartReceipt{}, err
	}
	t.state = TransactionStateStarted
	t.unlock()
	return receipt, nil
}

// Complete prepares input (generating and reserving a record_id on first
// call) and durably completes a previously started transaction, with
// result server-set to SUCCESS. It retries eligible transient failures
// with the exact same prepared bytes.
//
// Complete and Abort are mutually exclusive: once either has been
// chosen for this Transaction, calling the other returns an error
// without sending a request. Calling Complete before Record has
// succeeded, or after a terminal outcome already succeeded, is also
// rejected.
func (t *Transaction) Complete(ctx context.Context, input TransactionCompleteInput, opts RecordOptions) (TransactionTerminalReceipt, error) {
	return t.terminal(ctx, TransactionStateCompleted, transactionTerminalInput{
		RecordID:   input.RecordID,
		OccurredAt: input.OccurredAt,
		// Deliberately no Result here: the wire contract for /complete
		// has no result field at all (the server always sets it to
		// SUCCESS) — unlike /abort, which requires one. Setting Result
		// here would make buildTransactionTerminalObject include an
		// "unknown top-level field" the server's complete schema
		// rejects. The receipt's own Result still comes back as
		// "SUCCESS" from the server's response body.
		Outcome:  input.Outcome,
		Metadata: input.Metadata,
	}, opts)
}

// Abort prepares input and durably aborts a previously started
// transaction. Result defaults to ResultFailure when input.Result is
// empty; it must be ResultFailure or ResultDenied. See Complete for the
// shared mutual-exclusion and ordering rules.
func (t *Transaction) Abort(ctx context.Context, input TransactionAbortInput, opts RecordOptions) (TransactionTerminalReceipt, error) {
	result := input.Result
	if result == "" {
		result = ResultFailure
	}
	if result != ResultFailure && result != ResultDenied {
		return TransactionTerminalReceipt{}, newValidationError("abort result must be FAILURE or DENIED")
	}
	return t.terminal(ctx, TransactionStateAborted, transactionTerminalInput{
		RecordID:   input.RecordID,
		OccurredAt: input.OccurredAt,
		Result:     result,
		Outcome:    input.Outcome,
		Metadata:   input.Metadata,
	}, opts)
}

// terminal is the shared Complete/Abort implementation: kind is
// TransactionStateCompleted or TransactionStateAborted, chosen by the
// caller (Complete/Abort) before this function ever inspects state, so
// the mutual-exclusion check below always compares against the
// caller's actual intent.
func (t *Transaction) terminal(ctx context.Context, kind TransactionState, input transactionTerminalInput, opts RecordOptions) (TransactionTerminalReceipt, error) {
	t.lock()
	switch t.state {
	case TransactionStateNew, TransactionStateStartInFlight:
		t.unlock()
		return TransactionTerminalReceipt{}, newTransactionStateError("transaction: cannot complete/abort before Record has succeeded")
	case TransactionStateCompleteInFlight, TransactionStateAbortInFlight:
		t.unlock()
		return TransactionTerminalReceipt{}, newTransactionStateError("transaction: a call is already in flight; overlapping calls are not allowed")
	case TransactionStateCompleted, TransactionStateAborted:
		t.unlock()
		return TransactionTerminalReceipt{}, newTransactionStateError("transaction: this transaction already reached a terminal state")
	}
	if t.terminalChosen && t.terminalKind != kind {
		t.unlock()
		return TransactionTerminalReceipt{}, newTransactionStateError("transaction: complete and abort are mutually exclusive; this transaction already chose " + string(t.terminalKind))
	}
	// t.state == TransactionStateStarted here.
	if !t.terminalChosen {
		input.ReferencesRecordID = t.startRecordID
		prepared, err := createTransactionTerminal(input, t.client.now, t.client.random)
		if err != nil {
			t.unlock()
			return TransactionTerminalReceipt{}, err
		}
		t.terminalChosen = true
		t.terminalKind = kind
		t.terminalPrepared = prepared
	}
	prepared := t.terminalPrepared
	auditTransactionID := t.auditTransactionID
	inFlight := TransactionStateCompleteInFlight
	if kind == TransactionStateAborted {
		inFlight = TransactionStateAbortInFlight
	}
	t.state = inFlight
	t.unlock()

	receipt, err := t.client.sendTransactionTerminal(ctx, auditTransactionID, kind, prepared, opts)

	t.lock()
	if err != nil {
		// Revert to STARTED (not a terminal failure state) so a
		// caller-driven retry can call the same method again — the
		// mutual-exclusion check above already pins terminalKind, so a
		// caller cannot flip to the other terminal kind after this.
		t.state = TransactionStateStarted
		t.unlock()
		return TransactionTerminalReceipt{}, err
	}
	t.state = kind
	t.unlock()
	return receipt, nil
}

File: auditlog-js/src/transaction.ts
// Transaction: a one-owner, process-local helper for one Audit
// Transaction lifecycle: NEW -> START_IN_FLIGHT -> STARTED, then STARTED
// -> COMPLETE_IN_FLIGHT -> COMPLETED or STARTED -> ABORT_IN_FLIGHT ->
// ABORTED. A busy guard (a synchronous state check before the first
// `await` in each method) rejects an overlapping call rather than
// racing two in-flight requests for the same transaction — JS's
// single-threaded execution model makes this check-then-set sequence
// atomic with respect to any other call on the same Transaction
// instance, without needing a real lock.
//
// A Transaction holds no durable state and cannot be resumed after a
// process restart. If the process crashes after record() succeeds but
// before complete()/abort() is called, the evidence honestly remains
// STARTED — see the README's "Transaction lifecycle" section and
// internal/testutils/testdata/contracts/sdk/v1/transactions/manifest.json's
// process_crash_recovery guidance.
import { AuditError } from "./errors";
import { createTransactionStart, createTransactionTerminal } from "./transactionPrepare";
import type { PreparedTransactionStart, PreparedTransactionTerminal } from "./transactionPrepare";
import { sendTransactionStart, sendTransactionTerminal } from "./transactionRetry";
import type { SendTransactionDeps } from "./transactionRetry";
import type {
    AuditTransactionAbortInput,
    AuditTransactionCompleteInput,
    AuditTransactionStartInput,
    RequestOptions,
    TransactionStartReceipt,
    TransactionState,
    TransactionTerminalReceipt,
} from "./types";

export interface TransactionDeps extends SendTransactionDeps {
    now: () => Date;
    random: () => number;
}

export class Transaction {
    private readonly deps: TransactionDeps;
    private state: TransactionState = "NEW";

    private reserved = false;
    private auditTransactionId = "";
    private eventId = "";
    private startRecordId = "";
    private startPrepared: PreparedTransactionStart | undefined;

    private terminalChosen = false;
    private terminalKind: "COMPLETED" | "ABORTED" | undefined;
    private terminalPrepared: PreparedTransactionTerminal | undefined;

    constructor(deps: TransactionDeps) {
        this.deps = deps;
    }

    /** The Transaction's current state. */
    get currentState(): TransactionState {
        return this.state;
    }

    /** The reserved transaction ID, or "" before `record` has reserved one. */
    get auditTransactionIdValue(): string {
        return this.auditTransactionId;
    }

    /** The reserved event ID, or "" before `record` has reserved one. */
    get eventIdValue(): string {
        return this.eventId;
    }

    /** The reserved start record ID, or "" before `record` has reserved one. */
    get startRecordIdValue(): string {
        return this.startRecordId;
    }

    /**
     * Prepares `input` (generating and reserving audit_transaction_id/
     * event_id/record_id on first call) and durably starts the
     * transaction, retrying eligible transient failures with the exact
     * same prepared bytes. Resolves for both ACCEPTED and DUPLICATE.
     *
     * Calling `record` again after it already succeeded is rejected — a
     * Transaction starts at most once. Calling it again after a failed
     * attempt is allowed and reuses the exact same reserved IDs and
     * prepared bytes.
     */
    async record(input: AuditTransactionStartInput, options: RequestOptions = {}): Promise<TransactionStartReceipt> {
        if (this.state === "START_IN_FLIGHT" || this.state === "COMPLETE_IN_FLIGHT" || this.state === "ABORT_IN_FLIGHT") {
            throw AuditError.validation("transaction: a call is already in flight; overlapping calls are not allowed");
        }
        if (this.state !== "NEW") {
            throw AuditError.validation("transaction: record can only succeed once per transaction; this transaction has already started");
        }

        if (!this.reserved) {
            const prepared = createTransactionStart(input, { now: this.deps.now, random: this.deps.random });
            this.reserved = true;
            this.auditTransactionId = prepared.auditTransactionId;
            this.eventId = prepared.eventId;
            this.startRecordId = prepared.recordId;
            this.startPrepared = prepared;
        }
        const prepared = this.startPrepared as PreparedTransactionStart;
        this.state = "START_IN_FLIGHT";

        try {
            const receipt = await sendTransactionStart(prepared, options, this.deps);
            this.state = "STARTED";
            return receipt;
        } catch (error) {
            // Revert to NEW (not a terminal failure state) so a
            // caller-driven retry can call record again and reuse the
            // exact same reserved IDs and bytes reserved above.
            this.state = "NEW";
            throw error;
        }
    }

    /**
     * Prepares `input` (generating and reserving a record_id on first
     * call) and durably completes a previously started transaction, with
     * result server-set to SUCCESS. Deliberately omits `result` from the
     * request itself: the wire contract for complete has no result field
     * at all (unlike abort, which requires one) — sending one would make
     * the server reject the request as carrying an unknown top-level
     * field. The receipt's own `result` still comes back as `"SUCCESS"`
     * from the server's response body.
     */
    async complete(input: AuditTransactionCompleteInput, options: RequestOptions = {}): Promise<TransactionTerminalReceipt> {
        return this.terminal("COMPLETED", { record_id: input.record_id, occurred_at: input.occurred_at, outcome: input.outcome, metadata: input.metadata }, options);
    }

    /**
     * Prepares `input` and durably aborts a previously started
     * transaction. `result` defaults to `"FAILURE"` when omitted; it must
     * be `"FAILURE"` or `"DENIED"`. See `complete` for the shared
     * mutual-exclusion and ordering rules.
     */
    async abort(input: AuditTransactionAbortInput, options: RequestOptions = {}): Promise<TransactionTerminalReceipt> {
        const result = input.result ?? "FAILURE";
        if (result !== "FAILURE" && result !== "DENIED") {
            throw AuditError.validation("abort result must be FAILURE or DENIED");
        }
        return this.terminal("ABORTED", { record_id: input.record_id, occurred_at: input.occurred_at, result, outcome: input.outcome, metadata: input.metadata }, options);
    }

    private async terminal(
        kind: "COMPLETED" | "ABORTED",
        input: { record_id?: string; occurred_at: string | Date; result?: string; outcome?: AuditTransactionCompleteInput["outcome"]; metadata?: Record<string, unknown> },
        options: RequestOptions,
    ): Promise<TransactionTerminalReceipt> {
        if (this.state === "NEW" || this.state === "START_IN_FLIGHT") {
            throw AuditError.validation("transaction: cannot complete/abort before record has succeeded");
        }
        if (this.state === "COMPLETE_IN_FLIGHT" || this.state === "ABORT_IN_FLIGHT") {
            throw AuditError.validation("transaction: a call is already in flight; overlapping calls are not allowed");
        }
        if (this.state === "COMPLETED" || this.state === "ABORTED") {
            throw AuditError.validation("transaction: this transaction already reached a terminal state");
        }
        if (this.terminalChosen && this.terminalKind !== kind) {
            throw AuditError.validation(`transaction: complete and abort are mutually exclusive; this transaction already chose ${String(this.terminalKind)}`);
        }

        if (!this.terminalChosen) {
            const prepared = createTransactionTerminal(
                { record_id: input.record_id, references_record_id: this.startRecordId, occurred_at: input.occurred_at, result: input.result, outcome: input.outcome, metadata: input.metadata },
                { now: this.deps.now, random: this.deps.random },
            );
            this.terminalChosen = true;
            this.terminalKind = kind;
            this.terminalPrepared = prepared;
        }
        const prepared = this.terminalPrepared as PreparedTransactionTerminal;
        this.state = kind === "COMPLETED" ? "COMPLETE_IN_FLIGHT" : "ABORT_IN_FLIGHT";

        try {
            const receipt = await sendTransactionTerminal(this.auditTransactionId, kind === "COMPLETED" ? "complete" : "abort", prepared, options, this.deps);
            this.state = kind;
            return receipt;
        } catch (error) {
            // Revert to STARTED (not a terminal failure state) so a
            // caller-driven retry can call the same method again — the
            // mutual-exclusion check above already pins terminalKind, so
            // a caller cannot flip to the other terminal kind after this.
            this.state = "STARTED";
            throw error;
        }
    }
}
